### PR TITLE
Fixes #1623 Support classpath root nested in JAR

### LINKED ISSUE #1623: Support classpath root nested in JAR
This is the upstream issue for vert-x3/vertx-web#452 Allow templates to be loaded from classpath

Classpath root may be a folder inside a JAR, not the JAR itself.

For example, in Spring Boot, here's the list of root URLs:

```
0 = {URL@4064} "jar:file:/home/tsegismont/Temp/wv/target/testwb-1.0-SNAPSHOT.jar!/"
1 = {URL@4065} "jar:file:/home/tsegismont/Temp/wv/target/testwb-1.0-SNAPSHOT.jar!/BOOT-INF/classes!/"
2 = {URL@4066} "jar:file:/home/tsegismont/Temp/wv/target/testwb-1.0-SNAPSHOT.jar!/BOOT-INF/lib/vertx-web-templ-freemarker-3.4.0-SNAPSHOT.jar!/"
3 = {URL@4067} "jar:file:/home/tsegismont/Temp/wv/target/testwb-1.0-SNAPSHOT.jar!/BOOT-INF/lib/jackson-core-2.8.1.jar!/"
```

... and so on with all libs

Note the jar root, as well as BOOT-INF/classes

### PR DESCRIPTION
**Pull request by tsegismont**

### CODE REVIEW

File: src/main/java/io/vertx/core/impl/FileResolver.java [hunk 1]
```diff
@@ -182,12 +192,20 @@ private synchronized File unpackFromJarURL(URL url, String fileName, ClassLoader
         zip = new ZipFile(file);
       }
 
+      String inJarPath = path.substring(idx1 + 6);
+      String[] parts = JAR_URL_SEP_PATTERN.split(inJarPath);
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/FileResolver.java]: avoid String#split

File: src/main/java/io/vertx/core/impl/FileResolver.java [hunk 2]
```diff
@@ -182,12 +192,20 @@ private synchronized File unpackFromJarURL(URL url, String fileName, ClassLoader
...
+      }
+      String prefix = prefixBuilder.toString();
+
       Enumeration<? extends ZipEntry> entries = zip.entries();
       while (entries.hasMoreElements()) {
         ZipEntry entry = entries.nextElement();
         String name = entry.getName();
-        if (name.startsWith(fileName)) {
-          File file = new File(cacheDir, name);
+        if (name.startsWith(prefix + fileName)) {
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/FileResolver.java]: if prefix can be empty we should avoid string concat

**Comment by tsegismont** [comment]
Done

2016-09-21 7:05 GMT+02:00 Julien Viet <notifications@github.com>:

File: src/main/java/io/vertx/core/impl/FileResolver.java [hunk 3]
```diff
@@ -182,12 +192,20 @@ private synchronized File unpackFromJarURL(URL url, String fileName, ClassLoader
...
+      String prefix = prefixBuilder.toString();
+
       Enumeration<? extends ZipEntry> entries = zip.entries();
       while (entries.hasMoreElements()) {
         ZipEntry entry = entries.nextElement();
         String name = entry.getName();
-        if (name.startsWith(fileName)) {
-          File file = new File(cacheDir, name);
+        if (name.startsWith(prefix.isEmpty() ? fileName : prefix + fileName)) {
+          File file = new File(cacheDir, prefix.isEmpty() ? name : name.substring(prefix.length()));
```
Comment (vietj) [inline src/main/java/io/vertx/core/impl/FileResolver.java]: you can keep `name.substring(prefix.length)` : name.substring(0) returns `this`

**Comment by tsegismont** [comment]
I find this portion more readable with either both lines testing for empty
prefix or none. Would you mind if we keep it like this?

2016-09-21 14:54 GMT+02:00 Julien Viet <notifications@github.com>:
